api/client/inventory.go

```diff
@@ -0,0 +1,481 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/trace"
+	"github.com/gravitational/trace/trail"
+	log "github.com/sirupsen/logrus"
+)
+
+// DownstreamInventoryControlStream is the client/agent side of a bidirectional stream established
+// between teleport instances and auth servers.
+type DownstreamInventoryControlStream interface {
+	// Send attempts to send an upstream message. An error returned from this
+	// method either indicates that the stream itself has failed, or that the
+	// supplied context was canceled.
+	Send(ctx context.Context, msg proto.UpstreamInventoryMessage) error
+	// Recv accesses the incoming/downstream message channel.
+	Recv() <-chan proto.DownstreamInventoryMessage
+	// Close closes the underlying stream.
+	Close(...error)
+	// Done signals that the stream has been closed.
+	Done() <-chan struct{}
+	// Error checks for any error associated with stream closure (returns `nil` if
+	// the stream is open, or if the the stream was closed explicitly via the Close()
+	// method).
+	Error() error
+}
+
+// UpstreamInventoryControlStream is the server/controller side of a bidirectional stream established
+// between teleport instances and auth servers.
+type UpstreamInventoryControlStream interface {
+	// Send attempts to send a downstream message.  An error returned from this
+	// method either indicates that the stream itself has failed, or that the
+	// supplied context was canceled.
+	Send(ctx context.Context, msg proto.DownstreamInventoryMessage) error
+	// Recv access the incoming/upstream message channel.
+	Recv() <-chan proto.UpstreamInventoryMessage
+	// Close closes the underlying stream with optional error.
+	Close(...error)
+	// Done signals that the stream has been closed.
+	Done() <-chan struct{}
+	// Error checks for any error associated with stream closure (returns `nil` if
+	// the stream is open, or if the stream was closed explicitly via the Close()
+	// method).
+	Error() error
+}
+
+// InventoryControlStreamPipe creates the two halves of an inventory control stream over an in-memory
+// pipe.
+func InventoryControlStreamPipe() (UpstreamInventoryControlStream, DownstreamInventoryControlStream) {
+	pipe := &pipeControlStream{
+		downC: make(chan proto.DownstreamInventoryMessage, 1),
```

Comment:
I'm always suspicious of a buffered channel of size 1 - can you elaborate on why this is the case?

Reply:
Was using a buffer size of 1 here to keep the amount of slop as low as possible while still supporting simultaneous sending of "hello" messages (both sides of the stream were synchronously sending their own hello message out before trying to recv the incoming hello). Changed to having auth wait for client hello before sending its reply. Pros and cons to doing it either way, but above channels are now unbuffered.